src/cpu_shader/util.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2023 The Vello authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! Utility types
+
+#[derive(Clone, Copy, Default, Debug)]
+#[repr(C)]
+pub struct Vec2 {
+    pub x: f32,
+    pub y: f32,
+}
+
+impl std::ops::Add for Vec2 {
+    type Output = Vec2;
+
+    fn add(self, rhs: Self) -> Self {
+        Vec2 {
+            x: self.x + rhs.x,
+            y: self.y + rhs.y,
+        }
+    }
+}
+
+impl std::ops::Sub for Vec2 {
+    type Output = Vec2;
+
+    fn sub(self, rhs: Self) -> Self {
+        Vec2 {
+            x: self.x - rhs.x,
+            y: self.y - rhs.y,
+        }
+    }
+}
+
+impl std::ops::Mul<f32> for Vec2 {
+    type Output = Vec2;
+
+    fn mul(self, rhs: f32) -> Self {
+        Vec2 {
+            x: self.x * rhs,
+            y: self.y * rhs,
+        }
+    }
+}
+
+impl Vec2 {
+    pub fn new(x: f32, y: f32) -> Self {
+        Vec2 { x, y }
+    }
+
+    pub fn dot(self, other: Vec2) -> f32 {
+        self.x * other.x + self.y * other.y
+    }
+
+    pub fn length(self) -> f32 {
+        self.x.hypot(self.y)
+    }
+
+    pub fn to_array(self) -> [f32; 2] {
+        [self.x, self.y]
+    }
+
+    pub fn from_array(a: [f32; 2]) -> Self {
+        Vec2 { x: a[0], y: a[1] }
+    }
+
+    pub fn mix(self, other: Vec2, t: f32) -> Self {
+        let x = self.x + (other.x - self.x) * t;
+        let y = self.y + (other.y - self.y) * t;
+        Vec2 { x, y }
+    }
+}
+
+pub struct Transform(pub [f32; 6]);
+
+impl Transform {
+    pub fn apply(&self, p: Vec2) -> Vec2 {
+        let z = self.0;
+        let x = z[0] * p.x + z[2] * p.y + z[4];
```

Comment:
I have *not* checked these operations, but am reminded of https://github.com/linebender/vello/pull/283